### PR TITLE
OV-382-INITIALIZE-CLASS-TEST-CASES

### LINKED ISSUE #382: Initializer - Unit Tests
Full Unit Test suit needed for Initializer.scala

### PR DESCRIPTION
**Pull request by saswatdb**
closes #382

### CODE REVIEW

File: src/main/scala/com/databricks/labs/overwatch/pipeline/Initializer.scala [hunk 1]
```diff
@@ -77,8 +77,12 @@ class Initializer(config: Config) extends SparkSessionWrapper {
     if (config.consumerDatabaseName != config.databaseName) {
       logger.log(Level.INFO, "Initializing Consumer Database")
       if (!spark.catalog.databaseExists(config.consumerDatabaseName)) {
-        val createConsumerDBSTMT = s"create database if not exists ${config.consumerDatabaseName} " +
-          s"location '${config.consumerDatabaseLocation}'"
+        val createConsumerDBSTMT = if (!config.isLocalTesting) {
```
Comment (GeekSheikh) [inline src/main/scala/com/databricks/labs/overwatch/pipeline/Initializer.scala:80]: can we not set the location as a relative path in the config for local testing and not need a different logic branch for this?

Comment (GeekSheikh) [inline src/main/scala/com/databricks/labs/overwatch/pipeline/Initializer.scala:80]: can we not set the location as a relative path in the config for local testing and not need a different logic branch for this?

File: src/test/scala/com/databricks/labs/overwatch/pipeline/InitializeTest.scala [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+      val conf = new Config
+      conf.setOrganizationId("demo")
+      val init = new Initializer(conf)
+
+      val isPVC = PrivateMethod[Boolean]('isPVC)
+      val actual = init invokePrivate isPVC()
+      println("value: " + actual)
+      assert(!actual)
+    }
+  }
```
Comment (GeekSheikh) [inline src/test/scala/com/databricks/labs/overwatch/pipeline/InitializeTest.scala:24]: I think we should add a positive test case as well. i.e. where orgID has ilb in it

Comment (GeekSheikh) [inline src/test/scala/com/databricks/labs/overwatch/pipeline/InitializeTest.scala:24]: I think we should add a positive test case as well. i.e. where orgID has ilb in it

File: src/test/scala/com/databricks/labs/overwatch/pipeline/InitializeTest.scala [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+  describe("Tests for initialize database") {
+    it("initializeDatabase function should create both elt and consumer database") {
+      import spark.implicits._
+      val conf = new Config
+      conf.setDatabaseNameAndLoc("overwatch_etl", "/dblocation", "/datalocation")
+      conf.setConsumerDatabaseNameandLoc("overwatch", "/consumer/dblocation")
+      val init = new Initializer(conf)
+      val database = PrivateMethod[Database]('initializeDatabase)
+      init invokePrivate database()
+      val databases = spark.sql("show databases").select("namespace").map(f => f.getString(0)).collect()
```
Comment (GeekSheikh) [inline src/test/scala/com/databricks/labs/overwatch/pipeline/InitializeTest.scala:35]: what is this `select("namespace")`? It doesn't work on Databricks, not sure what you're getting here. I think `show databases` returns a column named `databaseName`

Comment (GeekSheikh) [inline src/test/scala/com/databricks/labs/overwatch/pipeline/InitializeTest.scala:35]: what is this `select("namespace")`? It doesn't work on Databricks, not sure what you're getting here. I think `show databases` returns a column named `databaseName`

File: src/test/scala/com/databricks/labs/overwatch/pipeline/InitializeTest.scala [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+      conf.setCloudProvider("aws")
+      val init = new Initializer(conf)
+      val quickBuildAuditLogConfig = PrivateMethod[AuditLogConfig]('validateAuditLogConfigs)
+
+      assertThrows[BadConfigException](init invokePrivate quickBuildAuditLogConfig(configInput))
+    }
+  }
+
+
+  describe("Tests for validateAndRegisterArgs function") {
```
Comment (GeekSheikh) [inline src/test/scala/com/databricks/labs/overwatch/pipeline/InitializeTest.scala:152]: Some additional tests are needed.

- test for missing parameters
- test for duplicate parameters
- test for non-escaped json
- test for malformed json

File: src/test/scala/com/databricks/labs/overwatch/pipeline/InitializeTest.scala [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+      val dataTarget = DataTarget(Some("overwatch_etl"),Some("/path/to/database"), Some("/path/prefix"), Some("overwatch"), Some("/path/to/consumer_database"))
+      spark.sql("create database if not exists overwatch_etl")
+
+      val conf = new Config
+      val init = new Initializer(conf)
+      val dataTargetIsValid = PrivateMethod[Boolean]('dataTargetIsValid)
+      assertThrows[BadConfigException](init invokePrivate dataTargetIsValid(dataTarget))
+    }
+
+    it("dataTargetIsValid function should throw exception when the current db is not created from overwatch") {
```
Comment (GeekSheikh) [inline src/test/scala/com/databricks/labs/overwatch/pipeline/InitializeTest.scala:186]: more tests:

- use initializeDatabase to create a database that is owned by overwatch and validate that it works as expected.
- additional tests around initDatabase and validate DataTarget for schema tests
- I don't see any tests for the metadata checks
- what if consumerDB is left blank in config -- does it work as expected (i.e. etl and consumer dbs are the same)?

I think there are likely more scenarios than this even...I want to make sure and have good coverage on `initDatabase` and `validateDataTarget`. Please feel free to ask some questions but I think if you review the code a little deeper you can find quite a few more cases.

File: src/test/scala/com/databricks/labs/overwatch/pipeline/InitializeTest.scala [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+  describe("Tests for validateScope function") {
+    import com.databricks.labs.overwatch.utils.DataTarget
+    it("validateScope function should throw exception when the variable in scope is not one of the valid scope") {
+
+      val conf = new Config
+      val init = new Initializer(conf)
+      val validateScope = PrivateMethod[Seq[OverwatchScope.OverwatchScope]]('validateScope)
+      assertThrows[BadConfigException](init invokePrivate validateScope(Seq("invalidScope")))
+    }
+    it("validateScope function should check if sparkEvents, clusterEvents, and jobs scopes require clusters scope") {
```
Comment (GeekSheikh) [inline src/test/scala/com/databricks/labs/overwatch/pipeline/InitializeTest.scala:205]: We need a lot more test cases here:

- I think I see 3 dependency chains here -- we should test them all independently 
- There's at least 1 more where all scopes are valid (positive test case)
- We need case sensitivity tests
- Seq[String] should maintain order, can we validate that? Are there any order dependencies that could break the logic chain?

File: src/main/scala/com/databricks/labs/overwatch/pipeline/Initializer.scala [hunk 7]
```diff
@@ -360,8 +364,8 @@ class Initializer(config: Config) extends SparkSessionWrapper {
     )
 
     // must happen AFTER data target validation
-    if (!disableValidations) { // temp working dir is not necessary for disabled validations as pipelines cannot be
-    // executed without validations
+    if (!disableValidations && !config.isLocalTesting) { // temp working dir is not necessary for disabled validations as pipelines cannot be
```
[the remark this answers is not in the document]
Comment (GeekSheikh) [reply src/main/scala/com/databricks/labs/overwatch/pipeline/Initializer.scala:364]: ah gotcha. I'm interested to know if we can get rid of these local testing checks eventually but for now this is good. Thank you.

**Review by GeekSheikh — COMMENTED** [review]
Thank you for the updates. The PR is looking good but now 4 tests are failing, can you fix these up?

```
- validateAuditLogConfigs function validate audit log format in the config  *** FAILED ***
[info]   Expected exception com.databricks.labs.overwatch.utils.BadConfigException to be thrown, but java.lang.NullPointerException was thrown (InitializeTest.scala:159)
[info] Tests for validateAndRegisterArgs function
[info] - validateAndRegisterArgs function should validate and register variables on the conf *** FAILED ***
[info]   java.lang.NullPointerException:
[info]   at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
[info]   at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
[info]   at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[info]   at java.lang.reflect.Method.invoke(Method.java:498)
[info]   at com.databricks.dbutils_v1.DBUtilsHolder$$anon$1.invoke(DBUtilsHolder.scala:17)
[info]   at com.sun.proxy.$Proxy18.secrets(Unknown Source)
[info]   at com.databricks.labs.overwatch.pipeline.Initializer.com$databricks$labs$overwatch$pipeline$Initializer$$validateAndRegisterArgs(Initializer.scala:313)
[info]   at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
[info]   at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
[info]   at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[info]   ...
[info] - validateAndRegisterArgs function should fail when a parameter is missing *** FAILED ***
```

**Comment by saswatdb** [comment]
> 

All 22 test cases getting passed for me, have we set these environment variables : 
OVERWATCH=LOCAL;OVERWATCH_ENV=demo;OVERWATCH_TOKEN=dapi6b1226434461814e7bh43

**Review by GeekSheikh — CHANGES_REQUESTED** [review]
@saswatdb There are 4 tests failing wrt to Initializer tests -- could you please fix these

**Comment by souravbaner-da** [comment]
Hi @GeekSheikh it seems all the test cases are running successfully after providing the env variables as below:
```
OVERWATCH_ENV=demo;
OVERWATCH_TOKEN=dapi6b1226434461814effff;
OVERWATCH=LOCAL
```

Do we have to provide the same during our deployment also as in Local all are working fine.

**Comment by souravbaner-da** [comment]
Attached is the screenshot for the same

image

**Comment by GeekSheikh** [comment]
Reminder - Add docs to tests folder README.md

Add details to building.md
Add reference and build summary with links in /readme.md

**Review by GeekSheikh — APPROVED** [review]
I would like to review a few of these ignored tests again later and see if we can't make them work outside of full pipeline but looks good approving.

**Pull request merged**